AbstractMethodError when executing CTE query with 1.6.10 and Hibernate ORM 6.4.0.Final

### Description
I updated our application to use Hibernate ORM 6.4.0.Final and tested a few things in our application. Executing a CTE query like the following leads to an exception. This is of course not our original query, but I cannot post that here.

```java
@NamedQuery(name = "Object.selectAllParentsInfoRecursive",
        query = "WITH ParentInfoCte AS(" +
                "SELECT p parent, c child, aa.id assignmentId, true directParent, aa.expiryDate expiryDate" +
                "  FROM ... aa " +
                " INNER JOIN aa.parent p" +
                " INNER JOIN aa.child c" +
                "  WHERE c.id = :childId" +
                "   AND aa.validFrom IS NOT NULL" +
                "   AND aa.validTo IS NULL " +
                "UNION ALL " +
                "SELECT p2 parent, c2 child, aa2.id assignmentId, false directParent, NULL expiryDate" +
                "  FROM ParentInfoCte pi " +
                " INNER JOIN ... aa2 on pi.parent = aa2.child " +
                " INNER JOIN aa2.parent p2" +
                " INNER JOIN aa2.child c2" +
                "  WHERE aa2.validFrom IS NOT NULL" +
                "   AND aa2.validTo IS NULL" +
                "   AND c2 != p2" +
                ") search breadth first by parent set orderAttr cycle assignmentId set cycleMark " +
                " SELECT new ParentInfo(pi.parent, pi.child, pi.directParent, pi.assignmentId, pi.expiryDate) " +
                "    FROM ParentInfoCte pi")
```

```
List<ParentInfo> resultList =
        em.createNamedQuery("Object.selectAllParentsInfoRecursive", ParentInfo.class)
                .setParameter("childId", adsObject.getId())
                .getResultList();
```

### Expected behavior
The query works and returns the desired results

### Actual behavior
The following exception occurs:
```
Caused by: java.lang.AbstractMethodError: Receiver class com.blazebit.persistence.integration.hibernate.base.function.HibernateJpqlFunctionAdapter does not define or inherit an implementation of the resolved method 'abstract void render(org.hibernate.sql.ast.spi.SqlAppender, java.util.List, org.hibernate.query.ReturnableType, org.hibernate.sql.ast.SqlAstTranslator)' of interface org.hibernate.query.sqm.function.FunctionRenderer.
    at org.hibernate@6.4.0.Final//org.hibernate.query.sqm.function.SelfRenderingFunctionSqlAstExpression.renderToSql(SelfRenderingFunctionSqlAstExpression.java:184)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitSelfRenderingExpression(AbstractSqlAstTranslator.java:6832)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.tree.expression.SelfRenderingExpression.accept(SelfRenderingExpression.java:20)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.render(AbstractSqlAstTranslator.java:6434)
    at org.hibernate@6.4.0.Final//org.hibernate.dialect.function.CastingConcatFunction.renderAsString(CastingConcatFunction.java:102)
    at org.hibernate@6.4.0.Final//org.hibernate.dialect.function.CastingConcatFunction.render(CastingConcatFunction.java:89)
    at org.hibernate@6.4.0.Final//org.hibernate.query.sqm.function.SelfRenderingFunctionSqlAstExpression.renderToSql(SelfRenderingFunctionSqlAstExpression.java:184)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitSelfRenderingExpression(AbstractSqlAstTranslator.java:6832)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.tree.expression.SelfRenderingExpression.accept(SelfRenderingExpression.java:20)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitRecursivePath(AbstractSqlAstTranslator.java:2500)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.emulateSearchClauseOrderWithString(AbstractSqlAstTranslator.java:2440)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.renderRecursiveCteVirtualSelections(AbstractSqlAstTranslator.java:2126)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.renderVirtualSelections(AbstractSqlAstTranslator.java:4837)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitSelectClause(AbstractSqlAstTranslator.java:4731)
    at org.hibernate@6.4.0.Final//org.hibernate.dialect.SQLServerSqlAstTranslator.visitSelectClause(SQLServerSqlAstTranslator.java:287)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitQuerySpec(AbstractSqlAstTranslator.java:3180)
    at org.hibernate@6.4.0.Final//org.hibernate.dialect.SQLServerSqlAstTranslator.visitQuerySpec(SQLServerSqlAstTranslator.java:277)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.tree.select.QuerySpec.accept(QuerySpec.java:119)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.renderQueryGroup(AbstractSqlAstTranslator.java:3105)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitQueryGroup(AbstractSqlAstTranslator.java:3025)
    at org.hibernate@6.4.0.Final//org.hibernate.dialect.SQLServerSqlAstTranslator.visitQueryGroup(SQLServerSqlAstTranslator.java:267)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.tree.select.QueryGroup.accept(QueryGroup.java:67)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitSelectStatement(AbstractSqlAstTranslator.java:998)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.tree.select.SelectStatement.accept(SelectStatement.java:68)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitCteDefinition(AbstractSqlAstTranslator.java:1933)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitCteStatement(AbstractSqlAstTranslator.java:1794)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitCteContainer(AbstractSqlAstTranslator.java:1701)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitSelectStatement(AbstractSqlAstTranslator.java:997)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.translateSelect(AbstractSqlAstTranslator.java:861)
    at org.hibernate@6.4.0.Final//org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.translate(AbstractSqlAstTranslator.java:811)
    at org.hibernate@6.4.0.Final//org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.buildCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:416)
    at org.hibernate@6.4.0.Final//org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:310)
    at org.hibernate@6.4.0.Final//org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performList(ConcreteSqmSelectQueryPlan.java:286)
    at org.hibernate@6.4.0.Final//org.hibernate.query.sqm.internal.QuerySqmImpl.doList(QuerySqmImpl.java:509)
    at org.hibernate@6.4.0.Final//org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:427)
    at org.hibernate@6.4.0.Final//org.hibernate.query.Query.getResultList(Query.java:120)

```

### Steps to reproduce
I think the problem is a simple incompatibility, so I did not create a reproducer.

### Environment
Version:            1.6.10
JPA-Provider:       Hibernate 6.4.0.Final
DBMS:               Microsoft SQL Server 2019
Application Server: WildFly 30

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ Changes that happened in releases
 
 * Fix over-fetching of entity view data with dynamic fetches
 * Fix Spring 6.1 compatibility
+* Fix Hibernate 6.4.0.Final compatibility
 
 ### Backwards-incompatible changes
 

--- a/integration/hibernate6-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/function/HibernateJpqlFunctionAdapter.java
+++ b/integration/hibernate6-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/function/HibernateJpqlFunctionAdapter.java
@@ -21,13 +21,13 @@ import com.blazebit.persistence.spi.JpqlFunctionKind;
 import org.hibernate.engine.spi.SessionFactoryImplementor;
 import org.hibernate.metamodel.mapping.BasicValuedMapping;
 import org.hibernate.metamodel.mapping.JdbcMappingContainer;
+import org.hibernate.metamodel.mapping.MappingModelExpressible;
 import org.hibernate.metamodel.model.domain.DomainType;
 import org.hibernate.metamodel.model.domain.PersistentAttribute;
 import org.hibernate.query.ReturnableType;
 import org.hibernate.query.sqm.SqmExpressible;
 import org.hibernate.query.sqm.function.AbstractSqmSelfRenderingFunctionDescriptor;
 import org.hibernate.query.sqm.function.FunctionKind;
-import org.hibernate.query.sqm.function.FunctionRenderingSupport;
 import org.hibernate.query.sqm.produce.function.FunctionReturnTypeResolver;
 import org.hibernate.query.sqm.produce.function.StandardFunctionArgumentTypeResolvers;
 import org.hibernate.query.sqm.tree.SqmTypedNode;
@@ -49,14 +49,24 @@ import java.util.function.Supplier;
  * @author Christian Beikov
  * @since 1.6.7
  */
-public class HibernateJpqlFunctionAdapter extends AbstractSqmSelfRenderingFunctionDescriptor implements FunctionRenderingSupport {
+public class HibernateJpqlFunctionAdapter extends AbstractSqmSelfRenderingFunctionDescriptor {
 
     private final JpqlFunction function;
 
     public HibernateJpqlFunctionAdapter(SessionFactoryImplementor sfi, JpqlFunctionKind kind, JpqlFunction function) {
         super(null, determineFunctionKind(kind), null, new FunctionReturnTypeResolver() {
-            @Override
-            public ReturnableType<?> resolveFunctionReturnType(ReturnableType<?> impliedType, List<? extends SqmTypedNode<?>> arguments, TypeConfiguration typeConfiguration) {
+            public ReturnableType<?> resolveFunctionReturnType(
+                    ReturnableType<?> impliedType,
+                    List<? extends SqmTypedNode<?>> arguments,
+                    TypeConfiguration typeConfiguration) {
+                return resolveFunctionReturnType(impliedType, null, arguments, typeConfiguration);
+            }
+
+            public ReturnableType<?> resolveFunctionReturnType(
+                    ReturnableType<?> impliedType,
+                    Supplier<MappingModelExpressible<?>> inferredTypeSupplier,
+                    List<? extends SqmTypedNode<?>> arguments,
+                    TypeConfiguration typeConfiguration) {
                 Class<?> argumentClass = null;
                 if (arguments.isEmpty()) {
                     Class<?> returnType = function.getReturnType(null);
@@ -131,15 +141,33 @@ public class HibernateJpqlFunctionAdapter extends AbstractSqmSelfRenderingFuncti
         return function;
     }
 
-    @Override
     public void render(SqlAppender sqlAppender, List<? extends SqlAstNode> sqlAstArguments, SqlAstTranslator<?> walker) {
-        render(sqlAppender, sqlAstArguments, null, walker);
+        render( sqlAppender, sqlAstArguments, null, (ReturnableType<?>) null, walker);
     }
 
-    @Override
-    public void render(SqlAppender sqlAppender, List<? extends SqlAstNode> sqlAstArguments, Predicate filter, SqlAstTranslator<?> walker) {
-        AbstractSqlAstTranslator<?> sqlAstTranslator = (AbstractSqlAstTranslator<?>) walker;
-        final StringBuilder sqlBuffer = sqlAstTranslator.getSqlBuffer();
+    public void render(
+            SqlAppender sqlAppender,
+            List<? extends SqlAstNode> sqlAstArguments,
+            Predicate filter,
+            SqlAstTranslator<?> walker) {
+        render( sqlAppender, sqlAstArguments, filter, (ReturnableType<?>) null, walker );
+    }
+
+    public void render(
+            SqlAppender sqlAppender,
+            List<? extends SqlAstNode> sqlAstArguments,
+            ReturnableType<?> returnableType,
+            SqlAstTranslator<?> sqlAstTranslator) {
+        render( sqlAppender, sqlAstArguments, null, returnableType, sqlAstTranslator );
+    }
+
+    public void render(
+            SqlAppender sqlAppender,
+            List<? extends SqlAstNode> sqlAstArguments,
+            Predicate filter,
+            ReturnableType<?> returnableType,
+            SqlAstTranslator<?> sqlAstTranslator) {
+        final StringBuilder sqlBuffer = ((AbstractSqlAstTranslator<?>) sqlAstTranslator).getSqlBuffer();
         List<String> sqlArguments = new ArrayList<>(sqlAstArguments.size());
         int startLength = sqlBuffer.length();
         for (SqlAstNode sqlAstArgument : sqlAstArguments) {

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -76,7 +76,7 @@
         <version.hibernate-5.6>5.6.10.Final</version.hibernate-5.6>
         <version.hibernate-6.2>6.2.13.Final</version.hibernate-6.2>
         <version.hibernate-6.3>6.3.1.Final</version.hibernate-6.3>
-        <version.hibernate-6.4>6.4.0.CR1</version.hibernate-6.4>
+        <version.hibernate-6.4>6.4.1.Final</version.hibernate-6.4>
 
         <version.datanucleus.base-4>4.1.17</version.datanucleus.base-4>
         <version.datanucleus.rdbms-4>4.1.19</version.datanucleus.rdbms-4>

